### PR TITLE
Creates OARS datasets, charts, and dashboards in Superset

### PR DESCRIPTION
**Pull request by pomegranited**
Finalizes Clickhouse + MySQL + Superset integration for https://github.com/openedx/openedx-oars/issues/30:
* Sets up an admin user to access the REST API
* Imports the exported Superset assets using `superset-api-client`, including database passwords for Clickhouse and MySQL.
* Publishes the imported dashboard and charts so any logged-in user with course access can see it.
* Adds the Row-Level Security Filters for the Clickhouse dataset during initialization.

**Dependencies**

* https://github.com/openedx/tutor-contrib-clickhouse/pull/4
* https://github.com/openedx/tutor-contrib-superset/pull/6

**Author Notes**

1. UUIDs for the Superset assets are all hard-coded in the `data/assets` dir. Should we make these configurable, in case they happen to clash with existing UUIDs?

**Testing instructions**

Set up OARS:

```
pip install git+https://github.com/openedx/tutor-contrib-clickhouse@main#egg=tutor_contrib_clickhouse
pip install git+https://git@github.com/openedx/tutor-contrib-oars@jill/superset-init-dashboard#egg=tutor_contrib_oars
pip install git+https://git@github.com/openedx/tutor-contrib-superset@jill/superset-xapi-permissions#egg=tutor_contrib_superset

tutor plugins enable clickhouse
tutor plugins enable superset
tutor plugins enable oars

# Load xapi demo data, enough so we get several courses
tutor config save --set CLICKHOUSE_LOAD_DEMO_DATA=xapi
tutor config save --set CLICKHOUSE_LOAD_DEMO_XAPI_BATCHES=500

tutor dev do init

# Create some test users
tutor dev do createuser --staff --superuser super_user super_user@email.com 
tutor dev do createuser --staff staff_user staff_user@email.com 
tutor dev do createuser course_staff course_staff@email.com 
tutor dev do createuser normal_user normal_user@email.com 
```

Superset should be running on port `:8088`, e.g. http://local.overhang.io:8088

Login to Superset as the superuser
1. Visit the Dashboards list: http://local.overhang.io:8088/dashboard/list/
    Ensure that you can see all the courses and blocks from the xapi demo data, and all their blocks.
1. Ensure you can add/modify Charts and Dashboards, and add SQL Lab queries (these queries don't use the Row-Level Security Filters, so access is currently limited to Superusers),
1. Check the Settings -- you should be able to view and edit all of the Security pages (don't recommend disturbing anything).
   superuser_settings

Login to Studio as the global staff or superuser, e.g. http://studio.local.overhang.io:8001
1. Create a course with an org+name+run that matches one of the xapi demo courses you saw in Superset.
2. Add the "course staff" user as Course Staff to this course.

Login to Superset as the global staff user:
1. Visit the Dashboards list: http://local.overhang.io:8088/dashboard/list/
    Ensure that you can see all the courses and blocks from the xapi demo data, and all their blocks.
1. Ensure you can add/modify Charts and Dashboards, add Databases and Datasets, but you can't modify Databases or Datasets, or make SQL Lab raw queries.
1. Check the Settings
   staff_settings

Login to Superset as the course staff user:
1. Visit the Dashboards list: http://local.overhang.io:8088/dashboard/list/
    Ensure that you can see only the course(s) that you have staff access to, and the blocks that belong to those courses.
    Browse through to the charts to re-confirm.
  *Note*: If the user gets access to any new courses later, they need to logout of Superset and back in to see them.
1. Check the Settings are very basic
   user_settings

Login to Superset as the normal user:
1. Visit the Dashboards list: http://local.overhang.io:8088/dashboard/list/
    No Dashboards should be visible.
1. Visit the Charts list: http://local.overhang.io:8088/chart/list/
    No Charts should be visible.
1. Check the Settings are very basic
   user_settings

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by pomegranited** [comment]
FYI @bmtcril -- will ping you when I'm ready for review.

But do you have any thoughts about this?

> **Author Notes & Concerns**
> 2. When launching Tutor from scratch with the Superset, Clickhouse and OARS plugins enabled, the initialization sequence for the plugins doesn't work. This plugin requires that Superset and Clickhouse to have run their tutor init scripts first before this plugin can init. So we may have to reorganize these plugins. [see forum question](https://discuss.openedx.org/t/initialization-order-for-tutor-plugins/9401/1)

**Comment by bmtcril** [comment]
This is super exciting! I think there may be some changes we'll need to sync up on re: ClickHouse table names based on the Ralph work, but I'll ping you on that once I get those PRs ready / merged.

I pinged tCRIL eng about the Tutor question and I think Kyle proposed a good solution to move forward there!

File: tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml
Comment (pomegranited) [inline tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml:1]: Originally, we needed these `{% raw %} {% endraw %}` so that jinja filters weren't expanded by Tutor during init.

But since the row-level security seems to mean that we don't need to embed our `can_view_course` filter logic in the queries themselves, then we probably don't need these `{% raw %} {% endraw %}` statements bookending all of these yaml files anymore either.

But if we do, then we should write a make script for adding them to an export zip.

File: tutoroars/templates/oars/jobs/init/superset-add-admin.sh
Comment (pomegranited) [inline tutoroars/templates/oars/jobs/init/superset-add-admin.sh:25]: Do https://github.com/openedx/tutor-contrib-superset/pull/6/files#diff-5749faaf28779d5be58a5c45a42924f058c0ce27d6191054130e059c881fe81e belong here too?

File: tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml
Comment (pomegranited) [inline tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml:2]: @bmtcril 

> I think there may be some changes we'll need to sync up on re: ClickHouse table names based on the Ralph work, but I'll ping you on that once I get those PRs ready / merged.

Thanks! I started using your new `xapi_events_all_parsed` table -- is that the right one?

File: tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml
Comment (bmtcril) [reply tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml:1]: I'm not sure we'll be able to use ClickHouse row level security without giving each user a ClickHouse user, which is probably prohibitively complicated to manage, but I'm open to suggestions there.

File: tutoroars/templates/oars/jobs/init/superset-add-admin.sh
Comment (bmtcril) [reply tutoroars/templates/oars/jobs/init/superset-add-admin.sh:25]: My gut is that the security features to work with the LMS belong here, but the xapi task belongs in the OARS repo, since I think it's possible people could want to use superset against the MySQL tables without xAPI. Does that sound right to you?

File: tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml
Comment (bmtcril) [reply tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml:2]: That's the one! 🎉

File: tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml
Comment (pomegranited) [reply tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xapi_events_all_parsed.yaml:1]: [Superset supports Row Level Security](https://superset.apache.org/docs/security/) :)
I can attach a RowLevelSecurityFilter to the Clickhouse xapi dataset and filter based on the `course_id`s the user is allowed to see. I've been able to do this manually, just have to figure out how to add/update the filter programmatically, since there isn't an import/export or command-line API for it.

File: tutoroars/templates/oars/jobs/init/superset-init-security.sh
Comment (pomegranited) [inline tutoroars/templates/oars/jobs/init/superset-init-security.sh:62]: RLS needs to be applied to the Open edX role -- have to figure out how to get SqlAlchemy to let me do that!!

**Comment by pomegranited** [comment]
UGHHH @bmtcril -- I was so sure I'd have this PR ready for your review today.. everything was working swimmingly in my devstack until I tore it all down and rebuilt it from scratch to do a final check.. and now, I can't see any data in my superset charts!! I can see it in the SQL Lab, so it doesn't seem to be a connection issue..

But I have no idea what's happened. Has to be something in the inits..

Anyway, I'm sorry I don't have this ready yet! Will aim to have it ready for you first thing on your Monday.

**Comment by bmtcril** [comment]
@pomegranited no worries, thanks for the amazing work! I'm hoping to have the Ralph Tutor plugin available early next week as well so we should have all of the pieces in place to finish the glue work and start end-to-end testing!

**Comment by bmtcril** [comment]
For this point: "UUIDs for the Superset assets are all hard-coded in the data/assets dir. Should we make these configurable, in case they happen to clash with existing UUIDs?" I think that preserving the UUIDs is good, if people make local modifications to a dashboard I'd like to not overwrite it on them. Presuming that's the actual behavior or something we can force. :)

**Comment by pomegranited** [comment]
@bmtcril  I found the issue! Pushed here: https://github.com/openedx/tutor-contrib-superset/pull/6/commits/e2112a0d9bb372e10a9600475f9b79a5db9347b7

> I'm hoping to have the Ralph Tutor plugin available early next week as well so we should have all of the pieces in place to finish the glue work and start end-to-end testing!

Yay!!

File: tutoroars/plugin.py
Comment (bmtcril) [inline tutoroars/plugin.py:68]: Is it right that they're all the same priority?

**Review by bmtcril — APPROVED** [review]
This is an amazing amount of work, thanks so much! Like the other one, I think we can merge here and move the TODO out to a follow-on task to look into. This will unblock so much!

File: tutoroars/plugin.py
Comment (pomegranited) [reply tutoroars/plugin.py:68]: Ah, good point... I'm not sure whether the order the tasks are added in applies another level of sorting on top of the priority ranking, but there's no reason to be ambiguous here.

Because `superset-add-admin.sh` must create the Admin user before `superset-api-dashboard.sh` can use it.. and really, `superset-init-security.sh` should set up the row-level permissions before `clickhouse-demo-xapi-data.sh` adds any xAPI data.

Comment (pomegranited) [reply tutoroars/plugin.py:71]: Fixed with [7325b92](https://github.com/openedx/tutor-contrib-oars/pull/3/commits/7325b92b184d54ba6439831a3a32b4e861c02917) :)

**Comment by openedx-webhooks** [comment]
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/tutor-contrib-oars&entry.867055334=https://github.com/openedx/tutor-contrib-oars/pull/3&entry.1484655318=https://github.com/pomegranited&entry.752974735=2023-02-22+11:45&entry.1917517419=2023-03-07+03:43&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
